crates/bevy_animation/src/graph.rs

```diff
@@ -0,0 +1,467 @@
+//! The animation graph.
+
+#![allow(missing_docs)]
+
+use std::ops::{Index, IndexMut, Mul};
+
+use bevy_asset::{Assets, Handle};
+use bevy_core::Name;
+use bevy_derive::Deref;
+use bevy_ecs::{
+    component::Component,
+    entity::Entity,
+    query::Has,
+    reflect::ReflectComponent,
+    system::{Query, Res},
+};
+use bevy_hierarchy::{Children, Parent};
+use bevy_reflect::Reflect;
+use bevy_render::mesh::morph::MorphWeights;
+use bevy_time::Time;
+use bevy_transform::components::Transform;
+use petgraph::{
+    stable_graph::{NodeIndex, StableDiGraph},
+    visit::{IntoNodeReferences, NodeRef},
+    Direction,
+};
+
+use crate::{AnimationClip, AnimationPlayer, PlayingAnimation, RepeatAnimation};
+
+#[derive(Component, Reflect)]
+#[reflect(Component)]
+pub struct AnimationGraph {
+    graph: StableDiGraph<AnimationNode, ()>,
+    root: AnimationNodeIndex,
+}
+
+#[derive(Clone, Copy, Deref, Reflect)]
+pub struct AnimationWeight(f32);
+
+#[derive(Clone, Reflect)]
+pub struct AnimationNode {
+    animation: Option<PlayingAnimation>,
+    pub paused: bool,
+    pub weight: AnimationWeight,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Deref, Reflect)]
+pub struct AnimationNodeIndex(NodeIndex<u32>);
+
+struct AnimationContext<
+    'gr,
+    't,
+    'ac,
+    'p,
+    'ppq,
+    'ppa,
+    'ppb,
+    'ppc,
+    'cq,
+    'ca,
+    'cb,
+    'cc,
+    'nq,
+    'na,
+    'nb,
+    'nc,
+    'tq,
+    'ta,
+    'tb,
+    'tc,
+    'mwq,
+    'mwa,
+    'mwb,
+    'mwc,
```

Comment:
These lifetimes definitely need to be addressed before we merge this. This is a little ridiculous.

Reply:
Can these not share the same lifetime bound? See for example [here](https://github.com/mbrea-c/bevy_animation_graph/blob/master/crates/bevy_animation_graph/src/core/context/pass_context.rs#L23) and [here](https://github.com/mbrea-c/bevy_animation_graph/blob/master/crates/bevy_animation_graph/src/core/context/system_resources.rs#L12).

Reply:
I was able to appease the borrow-checker with the following:
```rust
struct AnimationContext<'l, 'w, 's, 'c> {
    animation_graph: &'l mut AnimationGraph,
    time: &'l Time,
    root_entity: Entity,
    animation_clips: &'l Assets<AnimationClip>,
    parent: Option<&'w Parent>,
    parents: &'l Query<
        'w,
        's,
        (
            Has<AnimationPlayer>,
            Has<AnimationGraph>,
            Option<&'c Parent>,
        ),
    >,
    children: &'l Query<'w, 's, &'c Children>,
    names: &'l Query<'w, 's, &'c Name>,
    transforms: &'l Query<'w, 's, &'c mut Transform>,
    morph_weights: &'l Query<'w, 's, &'c mut MorphWeights>,
}

pub fn evaluate_animation_graphs<'w>(
    time: Res<Time>,
    animation_clips: Res<Assets<AnimationClip>>,
    children: Query<&'w Children>,
    names: Query<&'w Name>,
    transforms: Query<&'w mut Transform>,
    morph_weights: Query<&'w mut MorphWeights>,
    parents: Query<(
        Has<AnimationPlayer>,
        Has<AnimationGraph>,
        Option<&'w Parent>,
    )>,
    mut animation_graphs: Query<(Entity, Option<&'w Parent>, &'w mut AnimationGraph)>,
)
```
I don't *think* it is possible to reduce this further. Not currently sure what `AnimationContext` is for but I trust the process.